addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/discovery/IndividualAddressDiscoveryService.java

```diff
@@ -0,0 +1,139 @@
+/**
+ * Copyright (c) 2014 openHAB UG (haftungsbeschraenkt) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.knx.discovery;
+
+import static org.openhab.binding.knx.KNXBindingConstants.ADDRESS;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
+
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.knx.KNXBindingConstants;
+import org.openhab.binding.knx.KNXBusListener;
+import org.openhab.binding.knx.handler.KNXBridgeBaseThingHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Sets;
+
+import tuwien.auto.calimero.GroupAddress;
+import tuwien.auto.calimero.IndividualAddress;
+
+/**
+ * The {@link IndividualAddressDiscoveryService} class provides a discovery
+ * mechanism for KNX Individual Addresses
+ *
+ * @author Karel Goderis - Initial contribution
+ */
+public class IndividualAddressDiscoveryService extends AbstractDiscoveryService implements KNXBusListener {
+
+    public final static Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Sets
+            .newHashSet(KNXBindingConstants.THING_TYPE_GENERIC);
+
+    private final static int SEARCH_TIME = 600;
+    private boolean searchOngoing = false;
+
+    private KNXBridgeBaseThingHandler bridgeHandler;
+    protected Set<AreaLine> seenNetworks = new CopyOnWriteArraySet<>();
+
+    protected Logger logger = LoggerFactory.getLogger(KNXBridgeBaseThingHandler.class);
+
+    public IndividualAddressDiscoveryService(KNXBridgeBaseThingHandler bridgeHandler) throws IllegalArgumentException {
+        super(SUPPORTED_THING_TYPES_UIDS, SEARCH_TIME, false);
+        this.bridgeHandler = bridgeHandler;
+        bridgeHandler.registerKNXBusListener(this);
+    }
+
+    public class AreaLine {
+        public int area;
+        public int line;
+
+        public AreaLine(int area, int line) {
+            this.area = area;
+            this.line = line;
+        }
+
+        @Override
+        public boolean equals(Object o) {
```

Comment:
this implementation of equal is not null safe. And this class is lacking a `hashcode()` method. Which is especially a problem as it is used within a set!